### PR TITLE
feat(Entity): add updateMany by predicate

### PR DESCRIPTION
**Pull request by timdeschryver**
This PR allows to edit entities based on a predicate.
See #672, in the issue it was mentioned to update the entities via multiple predicates - see my comment as why I don't think thats a good idea. If needed I can implement it that way tho.

I do got one question about the implementation of the sorted state adapter. 
Why does this have an own implementation? 
In my head the sorted adapter should call the unsorted adapter and do the sorting afterwards, but I must be missing someting :sweat_smile:

### CODE REVIEW

**Comment by MikeRyanDev** [comment]
Implementation seems reasonable. What are your opinions on this:

Instead of a predicate with a list of changes the signature is updated to use a filter/map function:

```ts
 // Name is bad but it illustrated the interface
interface Updater<T> {
  (entity: T): false | Change<T>;
}
function updateMany(updater: Updater<T>, state: EntityState<T>);
```

**Comment by tdeschryver** [comment]
I don't think I get the idea.
Would this be a `map` implementation where the user implements `map`?
For example: 
```
adapter.updateMany(
      entity => entity.id.startsWith('a') ? {title: 'New title'} : false, // maybe null or an empty object instead of false?
      withMany
);
```

**Comment by MikeRyanDev** [comment]
Yes, that's right.

**Comment by tdeschryver** [comment]
I must say your proposal feels a bit weird at first but after seeing it in action, it is better in my opinion.
I did change the signature to return a `Partial<T>`, this brings the flexibility to do a second change.
See the unsorted state test: `book => (book.id.startsWith('a') ? firstChange : secondChange),`
Thoughts?

If someone got a better name than `Change` please feel free to speak up :smile:

File: modules/entity/src/models.ts [hunk 1]
```diff
@@ -38,6 +38,10 @@ export type UpdateNum<T> = {
 
 export type Update<T> = UpdateStr<T> | UpdateNum<T>;
 
+export type Map<T> = {
```
Comment (MikeRyanDev) [inline modules/entity/src/models.ts]: Recommend renaming this to something more specific. Maybe `UpdateMap`?

File: modules/entity/src/sorted_state_adapter.ts [hunk 2]
```diff
@@ -72,8 +73,16 @@ export function createSortedStateAdapter<T>(selectId: any, sort: any): any {
...
 
   function updateManyMutably(updates: Update<T>[], state: R): DidMutate;
-  function updateManyMutably(updates: any[], state: any): DidMutate {
+  function updateManyMutably(map: Map<T>, state: R): DidMutate;
+  function updateManyMutably(updatesOrMap: any[] | any, state: any): DidMutate {
     const models: T[] = [];
+    const updates: Update<T>[] =
+      updatesOrMap instanceof Array
+        ? updatesOrMap
+        : state.ids.map((id: string | number) => ({
```
Comment (MikeRyanDev) [inline modules/entity/src/sorted_state_adapter.ts]: I think that creating an `Update` object for each entity means even entities that shouldn't change will lose referential equality. I think this `map` should be a `reduce` instead that shrinks the `ids` array into a smaller array of `Update`s for only changed entities.

File: modules/entity/src/models.ts (hunk 1 above)
Comment (tdeschryver) [reply modules/entity/src/models.ts]: LGTM!

File: modules/entity/src/sorted_state_adapter.ts [hunk 3]
```diff
@@ -72,8 +73,25 @@ export function createSortedStateAdapter<T>(selectId: any, sort: any): any {
...
+  function updateManyMutably(map: UpdateMap<T>, state: R): DidMutate;
+  function updateManyMutably(updatesOrMap: any[] | any, state: any): DidMutate {
     const models: T[] = [];
+    const updates: Update<T>[] =
+      updatesOrMap instanceof Array
+        ? updatesOrMap
+        : state.ids.reduce((changes: Update<T>[], id: string | number) => {
+            const change = updatesOrMap(state.entities[id]);
+            if (change && Object.keys(change).length > 0) {
+              return [
```
Comment (MikeRyanDev) [inline modules/entity/src/sorted_state_adapter.ts]: Recommend changing this line to:
```ts
changes.push({ id, changes: change });

return changes;
```

That way we don't allocate and deallocate an array for every item we are upserting.

**Comment by MikeRyanDev** [comment]
One minor change. Otherwise this is looking good to me!

File: modules/entity/src/sorted_state_adapter.ts (hunk 3 above)
Comment (tdeschryver) [reply modules/entity/src/sorted_state_adapter.ts]: Done, thanks for the feedback/reviews!

Comment (MikeRyanDev) [reply modules/entity/src/sorted_state_adapter.ts]: @tdeschryver Thanks for all of the hard work! We really appreciate it! 😄

**Comment by sandangel** [comment]
Hi, how about adding this feature to upsert too?

**Comment by brandonroberts** [comment]
@timdeschryver will you rebase this so it can land?

**Comment by timdeschryver** [comment]
✔️

**Comment by MikeRyanDev** [comment]
I wonder if we could just have `entityAdapter.map(...)` instead.

**Comment by timdeschryver** [comment]
That would make sense, since this is what we're actually doing.
But I think having a `updateMany` would have its value if we would change the API to the following ?

```ts
{
  predicate: entity => boolean,
  changes: Partial<Entity>
}
```

**Comment by timdeschryver** [comment]
Ha, now I see we started with this a API  😄 -  https://github.com/ngrx/platform/pull/907/commits/149042b47dc55d62a0e28c092079899e23de6824

**Comment by timdeschryver** [comment]
I've updated the PR to be a `map` function.
While doing this I also changed the signature to actually return an entity object.

Let me know your thoughts about this one and if I have to change something.
I could also submit a PR if you agree with https://github.com/ngrx/platform/pull/907#issuecomment-421244609.

File: modules/entity/src/models.ts [hunk 4]
```diff
@@ -38,6 +38,10 @@ export type UpdateNum<T> = {
 
 export type Update<T> = UpdateStr<T> | UpdateNum<T>;
 
+export type EntityMap<T> = {
```
Comment (brandonroberts) [inline modules/entity/src/models.ts]: Update type to match `export type ComparerStr<T> = (a: T, b: T) => string;`

**Comment by brandonroberts** [comment]
Minor nit, otherwise changes LGTM

**Comment by timdeschryver** [comment]
There we go 😄

**Comment by brandonroberts** [comment]
@timdeschryver rebase pls

File: modules/entity/src/index.ts [hunk 5]
```diff
@@ -1,2 +1,9 @@
 export { createEntityAdapter } from './create_adapter';
-export { Dictionary, EntityState, EntityAdapter, Update } from './models';
+export {
+  Dictionary,
+  EntityState,
+  EntityAdapter,
+  Update,
+  EntityMap,
```
Comment (timdeschryver) [inline modules/entity/src/index.ts]: I did also `EntityMap` and `Predicate` here.

File: modules/store/src/store.ts [hunk 6]
```diff
@@ -19,9 +19,6 @@ export class Store<T> extends Observable<T> implements Observer<Action> {
     this.source = state$;
   }
 
-  /**
```
Comment (brandonroberts) [inline modules/store/src/store.ts]: Revert this change because its in a separate PR

Comment (timdeschryver) [reply modules/store/src/store.ts]: Doh, I was using 2 VSCode instances and was wondering what was happening.
Sorry... 😅
